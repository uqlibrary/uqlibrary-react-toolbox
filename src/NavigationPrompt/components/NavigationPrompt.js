import React from 'react';
import PropTypes from 'prop-types';
import {withRouter} from 'react-router-dom';

/**
 * A replacement component for the react-router `Prompt`.
 * Allows for more flexible dialogs.
 *
 * https://gist.github.com/bummzack/a586533607ece482475e0c211790dd50
 */
export class NavigationPrompt extends React.Component {
    static propTypes = {
        when: PropTypes.bool.isRequired,
        children: PropTypes.func.isRequired,
        history: PropTypes.object
    };

    constructor(props) {
        super(props);
        this.state = {nextLocation: null};
        this.confirmationBox = null;
    }

    componentDidMount() {
        this.unblock = this.props.history.block((nextLocation) => {
            if (this.props.when) {
                this.setState({
                    nextLocation: nextLocation
                });
                this.confirmationBox.showConfirmation();
                return !this.props.when;
            }

            return this.props.when;
        });
    }

    componentWillUnmount() {
        this.unblock();
    }

    setNavigationConfirmation = (ref) => {
        this.confirmationBox = ref;
    };

    _onCancel = () => {
        this.setState({nextLocation: null});
    };

    _onConfirm = () => {
        this.navigateToNextLocation();
    };

    navigateToNextLocation = () => {
        this.unblock();
        this.props.history.push(this.state.nextLocation.pathname);
    };

    render() {
        return (
            <div>
                {this.props.children(this.setNavigationConfirmation, this._onConfirm, this._onCancel)}
            </div>
        );
    }
}

export default withRouter(NavigationPrompt);
