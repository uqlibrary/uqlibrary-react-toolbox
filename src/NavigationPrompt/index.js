export {default as NavigationDialogBox} from './components/NavigationDialogBox';
