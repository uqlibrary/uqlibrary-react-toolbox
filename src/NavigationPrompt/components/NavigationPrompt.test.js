import { mount } from 'enzyme';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import {NavigationPrompt} from './NavigationPrompt';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import PropTypes from 'prop-types';
import {Provider} from 'react-redux';
import Immutable from 'immutable';
import injectTapEventPlugin from 'react-tap-event-plugin';
import {ConfirmDialogBox} from '../../index';

const create = () => {
    const initialState = Immutable.Map();

    const store = {
        getState: jest.fn(() => (initialState)),
        dispatch: jest.fn(),
        subscribe: jest.fn()
    };
    const next = jest.fn();
    const invoke = (action) => thunk(store)(next)(action);
    return {store, next, invoke}
};

function setup({when, history, isShallow = true}) {
    const txt = {
        confirmationTitle: 'Confirmation',
        confirmationMessage: 'Are you sure?',
        cancelButtonLabel: 'No',
        confirmButtonLabel: 'Yes'
    };

    const props = {
        when: when || false,
        history: history || {}
    };

    if (!isShallow) {
        return mount(
            <Provider store={create().store}>
                <NavigationPrompt {...props}>
                    {
                        (setConfirmation, onConfirm, onCancel) => (
                            <ConfirmDialogBox
                                onRef={setConfirmation}
                                onAction={onConfirm}
                                onCancelAction={onCancel}
                                locale={txt}
                            />
                        )
                    }
                </NavigationPrompt>
            </Provider>, {
                context: {
                    muiTheme: getMuiTheme()
                },
                childContextTypes: {
                    muiTheme: PropTypes.object.isRequired
                }
            });
    }

    return shallow(<Provider store={create().store}><NavigationPrompt {...props} /></Provider>);
}

beforeAll(() => {
    injectTapEventPlugin();
});

describe('NavigationPrompt component', () => {
    it('should render', () => {
        const testFunction = jest.fn();

        const history = {
            block: testFunction
        };

        const wrapper = setup({isShallow: false, history: history});
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(testFunction).toBeCalled();
    });
});
