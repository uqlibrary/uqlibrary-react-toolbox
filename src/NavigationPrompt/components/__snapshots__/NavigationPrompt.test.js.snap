// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationPrompt component should render 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <NavigationPrompt
    history={
      Object {
        "block": [Function],
      }
    }
    when={false}
  >
    <div>
      <ConfirmDialogBox
        hideCancelButton={false}
        locale={
          Object {
            "cancelButtonLabel": "No",
            "confirmButtonLabel": "Yes",
            "confirmationMessage": "Are you sure?",
            "confirmationTitle": "Confirmation",
          }
        }
        onAction={[Function]}
        onCancelAction={[Function]}
        onRef={[Function]}
      >
        <Dialog
          actions={
            Array [
              <div
                className="columns dialog-actions ConfirmDialogBox-actions"
              >
                <div
                  className="column is-hidden-mobile"
                />
                <div
                  className="column is-narrow"
                >
                  <RaisedButton
                    className="ConfirmDialogBox-actions-cancel"
                    disabled={false}
                    fullWidth={true}
                    label="No"
                    labelPosition="after"
                    onTouchTap={[Function]}
                    primary={false}
                    secondary={false}
                  />
                </div>
                <div
                  className="column is-narrow"
                >
                  <RaisedButton
                    className="ConfirmDialogBox-actions-confirm"
                    disabled={false}
                    fullWidth={true}
                    keyboardFocused={true}
                    label="Yes"
                    labelPosition="after"
                    onTouchTap={[Function]}
                    primary={false}
                    secondary={true}
                  />
                </div>
                <div
                  className="is-clearfix"
                />
              </div>,
            ]
          }
          autoDetectWindowHeight={true}
          autoScrollBodyContent={false}
          modal={true}
          open={false}
          repositionOnUpdate={true}
          title="Confirmation"
        >
          <RenderToLayer
            open={true}
            render={[Function]}
            useLayerForClickAway={false}
          />
        </Dialog>
      </ConfirmDialogBox>
    </div>
  </NavigationPrompt>
</Provider>
`;
